extensions/amp-story/1.0/amp-story-embedded-component.js

```diff
@@ -534,30 +559,93 @@ export class AmpStoryEmbeddedComponent {
           const pageRect = this.componentPage_./*OK*/getBoundingClientRect();
 
-          const centeredTop = pageRect.height / 2 - targetRect.height / 2;
-          const centeredLeft = pageRect.width / 2 - targetRect.width / 2;
-
-          // Only account for offset from target to page borders. Since in
-          // desktop mode page is not at the borders of viewport.
-          const leftOffset = targetRect.left - pageRect.left;
-          const topOffset = targetRect.top - pageRect.top;
-
-          state.translateY = centeredTop - topOffset;
-          state.translateX = leftOffset - centeredLeft;
+          // Retrieve scale() previously set by prepareForAnimation() on the
+          // element.
+          this.scaleFactor_ =
+            parseFloat(getStyle(target, 'transform').split('scale(')[1]);
+          const fullScreenWidth = targetRect.width / this.scaleFactor_;
+          // Gap on the left of the element between full-screen size and
+          // current size.
+          const leftGap = (fullScreenWidth - targetRect.width) / 2;
+          // Distance from left of page to what will be the left of the
+          // element in full-screen.
+          const fullScreenLeft = targetRect.left - leftGap - pageRect.left;
+          const centeredLeft = pageRect.width / 2 - fullScreenWidth / 2;
+          state.translateX = centeredLeft - fullScreenLeft;
+
+          // Calculate what the height of the element will be in full-screen.
+          const fullScreenHeight = targetRect.height / this.scaleFactor_;
+          // Gap on the top of the element between full-screen size and
+          // current size.
+          const topGap = (fullScreenHeight - targetRect.height) / 2;
+          // Distance from top of page to what will be the top of the element in
+          // full-screen.
+          const fullScreenTop = targetRect.top - topGap - pageRect.top;
+          const centeredTop = pageRect.height / 2 - fullScreenHeight / 2;
+          state.translateY = centeredTop - fullScreenTop;
         },
         /** mutate */
         () => {
           target.classList.add('i-amphtml-animate-expand-in');
+          target.classList.toggle('i-amphtml-expanded-component', true);
           setImportantStyles(dev().assertElement(target),
               {
                 transform: `translate3d(${state.translateX}px,
-                    ${state.translateY}px, 0)`,
+                  ${state.translateY}px, 0) scale(1)`,
               });
         });
   }
 
+  /**
+   * Resizes expandable element before it is expanded to full-screen, in
+   * preparation for its animation. It resizes it to its full-screen size, and
+   * scales it down to match size set by publisher, adding negative margins so
+   * that content around stays put.
+   * @param {!Element} page
+   * @param {!Element} element
+   * @param {!../../../src/service/resources-impl.Resources} resources
+   */
+  static prepareForAnimation(page, element, resources) {
```

Comment:
This code needs to run again on resize. Up to you whether you want to do it in this PR or not, but you might have to change some code for it to work.

Reply:
I tried implementing for this use case but it looks like it's going to be trickier than just running this code again. Once an embedded component has been ran through `prepareForAnimation()`, it will have new dimensions + margins + it'll be scaled down. So when calling `getBoundingClientRect()` on the embed after this, it will get dimensions that do not reflect the actual size of the component.

Do you have any suggestions? 
 @gmajoulet  @newmuis